<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.7"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>ngsaipp: ngsai::KineticClassifier Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">ngsaipp
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.7 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacengsai.html">ngsai</a></li><li class="navelem"><a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classngsai_1_1KineticClassifier-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">ngsai::KineticClassifier Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="KineticClassifier_8hpp_source.html">KineticClassifier.hpp</a>&gt;</code></p>
<div class="dynheader">
Collaboration diagram for ngsai::KineticClassifier:</div>
<div class="dyncontent">
<div class="center"><img src="classngsai_1_1KineticClassifier__coll__graph.png" border="0" usemap="#angsai_1_1KineticClassifier_coll__map" alt="Collaboration graph"/></div>
<map name="angsai_1_1KineticClassifier_coll__map" id="angsai_1_1KineticClassifier_coll__map">
<area shape="rect" title=" " alt="" coords="5,108,161,133"/>
<area shape="rect" href="classngsai_1_1KineticModel.html" title="The KineticModel class is an abstract class providing an interface to derive for classes that impleme..." alt="" coords="15,5,151,31"/>
</map>
<center><span class="legend">[<a href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ae8386f27eeaef5436d4798185ef0cecf"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#ae8386f27eeaef5436d4798185ef0cecf">KineticClassifier</a> ()</td></tr>
<tr class="memdesc:ae8386f27eeaef5436d4798185ef0cecf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor.  <br /></td></tr>
<tr class="separator:ae8386f27eeaef5436d4798185ef0cecf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e4f2955522ebfabc9f3c674e1e20f58"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a5e4f2955522ebfabc9f3c674e1e20f58">KineticClassifier</a> (<a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *model_meth, <a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *model_unmeth)</td></tr>
<tr class="memdesc:a5e4f2955522ebfabc9f3c674e1e20f58"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructs a classifier using the given kinetic signal models for predicion. It takes ownership over the KineticModels.  <br /></td></tr>
<tr class="separator:a5e4f2955522ebfabc9f3c674e1e20f58"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3e5bd0469911703ba8b218a3340b310"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#ad3e5bd0469911703ba8b218a3340b310">KineticClassifier</a> (const <a class="el" href="classngsai_1_1KineticClassifier.html">ngsai::KineticClassifier</a> &amp;other)</td></tr>
<tr class="memdesc:ad3e5bd0469911703ba8b218a3340b310"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy constructor.  <br /></td></tr>
<tr class="separator:ad3e5bd0469911703ba8b218a3340b310"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54e93094a2486513b4eab74c4c8b2571"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a54e93094a2486513b4eab74c4c8b2571">KineticClassifier</a> (<a class="el" href="classngsai_1_1KineticClassifier.html">ngsai::KineticClassifier</a> &amp;&amp;other)</td></tr>
<tr class="memdesc:a54e93094a2486513b4eab74c4c8b2571"><td class="mdescLeft">&#160;</td><td class="mdescRight">Move constructor.  <br /></td></tr>
<tr class="separator:a54e93094a2486513b4eab74c4c8b2571"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa7b00d61d4e7a7eaa43a6345d6a3427a"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#aa7b00d61d4e7a7eaa43a6345d6a3427a">~KineticClassifier</a> ()</td></tr>
<tr class="memdesc:aa7b00d61d4e7a7eaa43a6345d6a3427a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor. It will free all <a class="el" href="classngsai_1_1KineticModel.html" title="The KineticModel class is an abstract class providing an interface to derive for classes that impleme...">KineticModel</a> memory!  <br /></td></tr>
<tr class="separator:aa7b00d61d4e7a7eaa43a6345d6a3427a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4282266803db2143c868916e614c6055"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a4282266803db2143c868916e614c6055">operator=</a> (const <a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp;other)</td></tr>
<tr class="memdesc:a4282266803db2143c868916e614c6055"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assignment operator.  <br /></td></tr>
<tr class="separator:a4282266803db2143c868916e614c6055"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3110804d9af6dcad148e8a15fe6e065"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#ad3110804d9af6dcad148e8a15fe6e065">operator=</a> (<a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp;&amp;other)</td></tr>
<tr class="memdesc:ad3110804d9af6dcad148e8a15fe6e065"><td class="mdescLeft">&#160;</td><td class="mdescRight">Move assignment operator.  <br /></td></tr>
<tr class="separator:ad3110804d9af6dcad148e8a15fe6e065"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4ce5d987acbd2c4754a11c982716ee1c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a4ce5d987acbd2c4754a11c982716ee1c">setModels</a> (<a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *model_meth, <a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *model_unmeth)</td></tr>
<tr class="memdesc:a4ce5d987acbd2c4754a11c982716ee1c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the classifier models.  <br /></td></tr>
<tr class="separator:a4ce5d987acbd2c4754a11c982716ee1c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3f29c9122e9afebd64da0603c56bd868"><td class="memItemLeft" align="right" valign="top">std::pair&lt; double, double &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a3f29c9122e9afebd64da0603c56bd868">classify</a> (const <a class="el" href="classngsai_1_1genome_1_1CpGRegion.html">ngsai::genome::CpGRegion</a> &amp;cpg, const std::list&lt; PacBio::BAM::BamRecord &gt; &amp;ccss, double prob_meth, double prob_unmeth) const</td></tr>
<tr class="memdesc:a3f29c9122e9afebd64da0603c56bd868"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes the posterior probabilities p(methylated | IPD,PWD) and p(unmethylated | IPD,PWD) for the given region, given the CCS read kinetics. The prior class probabilites p(methylated) and p(unmethylated) are normalized to sum up to 1.  <br /></td></tr>
<tr class="separator:a3f29c9122e9afebd64da0603c56bd868"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c005db7983a06a3e8418e353865fc15"><td class="memItemLeft" align="right" valign="top">std::pair&lt; double, double &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a7c005db7983a06a3e8418e353865fc15">classify</a> (const <a class="el" href="classngsai_1_1KineticSignal.html">ngsai::KineticSignal</a> &amp;kinetics, double prob_meth, double prob_unmeth) const</td></tr>
<tr class="memdesc:a7c005db7983a06a3e8418e353865fc15"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes the posterior probabilities p(methylated | IPD,PWD) and p(unmethylated | IPD,PWD) for the given window signal.  <br /></td></tr>
<tr class="separator:a7c005db7983a06a3e8418e353865fc15"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pro-methods" name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:a6c613b8c9aedb03a1e4b2e75f03a9a91"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a6c613b8c9aedb03a1e4b2e75f03a9a91">checkModels</a> () const</td></tr>
<tr class="memdesc:a6c613b8c9aedb03a1e4b2e75f03a9a91"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks that all model size are the same and that the models contain log probability densities, hence whether they can be used to compute likelihood.  <br /></td></tr>
<tr class="separator:a6c613b8c9aedb03a1e4b2e75f03a9a91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5b0e930141daa13866398e2e02e75215"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a5b0e930141daa13866398e2e02e75215">freeModels</a> ()</td></tr>
<tr class="memdesc:a5b0e930141daa13866398e2e02e75215"><td class="mdescLeft">&#160;</td><td class="mdescRight">Frees the model memory if the pointer point to something.  <br /></td></tr>
<tr class="separator:a5b0e930141daa13866398e2e02e75215"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pro-attribs" name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:a17c780e76d99053eb5686acc5813e332"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a17c780e76d99053eb5686acc5813e332">m_window_size</a></td></tr>
<tr class="memdesc:a17c780e76d99053eb5686acc5813e332"><td class="mdescLeft">&#160;</td><td class="mdescRight">The window size in bases.  <br /></td></tr>
<tr class="separator:a17c780e76d99053eb5686acc5813e332"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a111f7a1e70501679d04477c33bd3e073"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a111f7a1e70501679d04477c33bd3e073">m_model_meth</a></td></tr>
<tr class="memdesc:a111f7a1e70501679d04477c33bd3e073"><td class="mdescLeft">&#160;</td><td class="mdescRight">The methylated kinetic signal model.  <br /></td></tr>
<tr class="separator:a111f7a1e70501679d04477c33bd3e073"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5790ce6e71280084bd275ffab338bb1a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classngsai_1_1KineticClassifier.html#a5790ce6e71280084bd275ffab338bb1a">m_model_unmeth</a></td></tr>
<tr class="memdesc:a5790ce6e71280084bd275ffab338bb1a"><td class="mdescLeft">&#160;</td><td class="mdescRight">The unmethylated kinetic signal model.  <br /></td></tr>
<tr class="separator:a5790ce6e71280084bd275ffab338bb1a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>The <a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> class implements a baysian classifier that determine the methylation state of CpGs from PacBio CCS data. The classifier implements a 2 class model (methylated and unmethylated) and assigns a probability of belonging to each class, for each CpG, given the kinetics of the CCS mapping over the CpG region. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="ae8386f27eeaef5436d4798185ef0cecf" name="ae8386f27eeaef5436d4798185ef0cecf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae8386f27eeaef5436d4798185ef0cecf">&#9670;&#160;</a></span>KineticClassifier() <span class="overload">[1/4]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">ngsai::KineticClassifier::KineticClassifier </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor. </p>

</div>
</div>
<a id="a5e4f2955522ebfabc9f3c674e1e20f58" name="a5e4f2955522ebfabc9f3c674e1e20f58"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5e4f2955522ebfabc9f3c674e1e20f58">&#9670;&#160;</a></span>KineticClassifier() <span class="overload">[2/4]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">ngsai::KineticClassifier::KineticClassifier </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *&#160;</td>
          <td class="paramname"><em>model_meth</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *&#160;</td>
          <td class="paramname"><em>model_unmeth</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructs a classifier using the given kinetic signal models for predicion. It takes ownership over the KineticModels. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">model_meth</td><td>the model for the methylated kinetic signal. </td></tr>
    <tr><td class="paramname">model_unmeth</td><td>the model for the unmethylated kinetic signal. </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if the models don't have an identical size. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ad3e5bd0469911703ba8b218a3340b310" name="ad3e5bd0469911703ba8b218a3340b310"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad3e5bd0469911703ba8b218a3340b310">&#9670;&#160;</a></span>KineticClassifier() <span class="overload">[3/4]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">ngsai::KineticClassifier::KineticClassifier </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classngsai_1_1KineticClassifier.html">ngsai::KineticClassifier</a> &amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Copy constructor. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>an instance to copy from. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a54e93094a2486513b4eab74c4c8b2571" name="a54e93094a2486513b4eab74c4c8b2571"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a54e93094a2486513b4eab74c4c8b2571">&#9670;&#160;</a></span>KineticClassifier() <span class="overload">[4/4]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">ngsai::KineticClassifier::KineticClassifier </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classngsai_1_1KineticClassifier.html">ngsai::KineticClassifier</a> &amp;&amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Move constructor. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>a rvalue to the object to move. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="aa7b00d61d4e7a7eaa43a6345d6a3427a" name="aa7b00d61d4e7a7eaa43a6345d6a3427a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa7b00d61d4e7a7eaa43a6345d6a3427a">&#9670;&#160;</a></span>~KineticClassifier()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">ngsai::KineticClassifier::~KineticClassifier </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Destructor. It will free all <a class="el" href="classngsai_1_1KineticModel.html" title="The KineticModel class is an abstract class providing an interface to derive for classes that impleme...">KineticModel</a> memory! </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a6c613b8c9aedb03a1e4b2e75f03a9a91" name="a6c613b8c9aedb03a1e4b2e75f03a9a91"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6c613b8c9aedb03a1e4b2e75f03a9a91">&#9670;&#160;</a></span>checkModels()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void ngsai::KineticClassifier::checkModels </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Checks that all model size are the same and that the models contain log probability densities, hence whether they can be used to compute likelihood. </p>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::runtime_error</td><td>if this is not the case. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a3f29c9122e9afebd64da0603c56bd868" name="a3f29c9122e9afebd64da0603c56bd868"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3f29c9122e9afebd64da0603c56bd868">&#9670;&#160;</a></span>classify() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::pair&lt; double, double &gt; ngsai::KineticClassifier::classify </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classngsai_1_1genome_1_1CpGRegion.html">ngsai::genome::CpGRegion</a> &amp;&#160;</td>
          <td class="paramname"><em>cpg</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::list&lt; PacBio::BAM::BamRecord &gt; &amp;&#160;</td>
          <td class="paramname"><em>ccss</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>prob_meth</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>prob_unmeth</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Computes the posterior probabilities p(methylated | IPD,PWD) and p(unmethylated | IPD,PWD) for the given region, given the CCS read kinetics. The prior class probabilites p(methylated) and p(unmethylated) are normalized to sum up to 1. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cpg</td><td>the coordinates of a CpG to classify, in genome coordinates. The CpG on both strand will be considered. The coordinates must be strandless. Start must be correspond to the coordinate of the C on the + strand and end to passed the G coordinate on the + strand. </td></tr>
    <tr><td class="paramname">ccss</td><td>a list of CCS reads mapping on the CpG of interest. Their IPD and PWD kinetics will be used to compute the probability that the CpG is methylated or not methylated. CCS not mapping over the CpG or without a perfect alignment will be ignored. </td></tr>
    <tr><td class="paramname">prob_meth</td><td>the a priori probability of the methylated class. </td></tr>
    <tr><td class="paramname">prob_unmeth</td><td>the a priori probability of the unmethylated class. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a pair of posterior probability p(methylated | IPD,PWD) and p(unmethylated | IPD,PWD) respectively. The probabilities sum to 1. </dd></dl>

</div>
</div>
<a id="a7c005db7983a06a3e8418e353865fc15" name="a7c005db7983a06a3e8418e353865fc15"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7c005db7983a06a3e8418e353865fc15">&#9670;&#160;</a></span>classify() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::pair&lt; double, double &gt; ngsai::KineticClassifier::classify </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classngsai_1_1KineticSignal.html">ngsai::KineticSignal</a> &amp;&#160;</td>
          <td class="paramname"><em>kinetics</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>prob_meth</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>prob_unmeth</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Computes the posterior probabilities p(methylated | IPD,PWD) and p(unmethylated | IPD,PWD) for the given window signal. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">kinetics</td><td>an instance containing the kinetics (IPD and PWD) and DNA sequences on the forward and reverse strand of the region to classify. </td></tr>
    <tr><td class="paramname">prob_meth</td><td>the a priori probability of the methylated class. </td></tr>
    <tr><td class="paramname">prob_unmeth</td><td>the a priori probability of the unmethylated class. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a pair of posterior probability p(methylated | IPD,PWD) and p(unmethylated | IPD,PWD) respectively. The probabilities sum to 1. </dd></dl>

</div>
</div>
<a id="a5b0e930141daa13866398e2e02e75215" name="a5b0e930141daa13866398e2e02e75215"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5b0e930141daa13866398e2e02e75215">&#9670;&#160;</a></span>freeModels()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void ngsai::KineticClassifier::freeModels </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Frees the model memory if the pointer point to something. </p>

</div>
</div>
<a id="a4282266803db2143c868916e614c6055" name="a4282266803db2143c868916e614c6055"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4282266803db2143c868916e614c6055">&#9670;&#160;</a></span>operator=() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp; ngsai::KineticClassifier::operator= </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Assignment operator. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>an instance to assign from. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to the instance. </dd></dl>

</div>
</div>
<a id="ad3110804d9af6dcad148e8a15fe6e065" name="ad3110804d9af6dcad148e8a15fe6e065"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad3110804d9af6dcad148e8a15fe6e065">&#9670;&#160;</a></span>operator=() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp; ngsai::KineticClassifier::operator= </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classngsai_1_1KineticClassifier.html">KineticClassifier</a> &amp;&amp;&#160;</td>
          <td class="paramname"><em>other</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Move assignment operator. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">other</td><td>an instance to assign from. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to the instance. </dd></dl>

</div>
</div>
<a id="a4ce5d987acbd2c4754a11c982716ee1c" name="a4ce5d987acbd2c4754a11c982716ee1c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4ce5d987acbd2c4754a11c982716ee1c">&#9670;&#160;</a></span>setModels()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void ngsai::KineticClassifier::setModels </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *&#160;</td>
          <td class="paramname"><em>model_meth</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a> *&#160;</td>
          <td class="paramname"><em>model_unmeth</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the classifier models. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">model_meth</td><td>the signal model for the methylated kinetics. </td></tr>
    <tr><td class="paramname">model_unmeth</td><td>the signal model for the unmethylated kinetics. </td></tr>
  </table>
  </dd>
</dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">std::invalid_argument</td><td>if all models are not initialised or if all models don't have the same size. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a111f7a1e70501679d04477c33bd3e073" name="a111f7a1e70501679d04477c33bd3e073"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a111f7a1e70501679d04477c33bd3e073">&#9670;&#160;</a></span>m_model_meth</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a>* ngsai::KineticClassifier::m_model_meth</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The methylated kinetic signal model. </p>

</div>
</div>
<a id="a5790ce6e71280084bd275ffab338bb1a" name="a5790ce6e71280084bd275ffab338bb1a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5790ce6e71280084bd275ffab338bb1a">&#9670;&#160;</a></span>m_model_unmeth</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classngsai_1_1KineticModel.html">ngsai::KineticModel</a>* ngsai::KineticClassifier::m_model_unmeth</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The unmethylated kinetic signal model. </p>

</div>
</div>
<a id="a17c780e76d99053eb5686acc5813e332" name="a17c780e76d99053eb5686acc5813e332"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a17c780e76d99053eb5686acc5813e332">&#9670;&#160;</a></span>m_window_size</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t ngsai::KineticClassifier::m_window_size</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The window size in bases. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>include/ngsaipp/epigenetics/<a class="el" href="KineticClassifier_8hpp_source.html">KineticClassifier.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.7
</small></address>
</body>
</html>
